lib/CommandClient.php

```diff
@@ -26,22 +32,82 @@ public static function socketPath(string $config) {
     private function send($message): Promise {
         return call(function () use ($message) {
             if (!$this->socket) {
-                $this->socket = yield from $this->establish();
+                yield from $this->establish();
             }
 
-            $message = \json_encode($message);
-            $message = \pack("N", \strlen($message)) . $message;
-            return yield $this->socket->write($message);
+            try {
+                $message = \json_encode($message) . "\n";
+                yield $this->socket->write($message);
+            } catch (ClosedException $e) {
+                yield from $this->establish();
+                yield $this->socket->write($message);
+            }
+
+            $promise = ($this->readDeferreds[] = new Deferred)->promise();
+
+            // we can ONLY resolve the socketDeferred AFTER the first write, otherwise order of send()'s is not respected
```

Comment:
Do we really need that? Registered `onResolve` callbacks should be executed in order.